lib/ide/src/transport/web.rs

```diff
@@ -0,0 +1,250 @@
+    pub on_error   : ClosureStorage<Event>,
+}
+
+impl WebSocket {
+    /// Wraps given WebSocket object.
+    pub fn new(ws:web_sys::WebSocket) -> WebSocket {
+        WebSocket {
+            ws,
+            on_message : default(),
+            on_close   : default(),
+            on_open    : default(),
+            on_error   : default(),
+        }
+    }
+
+    /// Establish connection with endpoint defined by the given URL and wrap it.
+    /// Asynchronous, because it waits until connection is established.
+    pub async fn new_opened(url:impl Str) -> Result<WebSocket,ConnectingError> {
+        let     ws  = web_sys::WebSocket::new(url.as_ref());
+        let mut wst = WebSocket::new(ws.map_err(|e| {
+            ConnectingError::ConstructionError(js_to_string(e))
+        })?);
+
+        wst.wait_until_open().await?;
+        Ok(wst)
+    }
+
+    /// Awaits until `open` signal has been emitted. Clears any callbacks on
+    /// this `WebSocket`, if any has been set.
+    async fn wait_until_open(&mut self) -> Result<(),ConnectingError> {
+        // Connecting attempt shall either emit on_open or on_close.
+        // We shall wait for whatever comes first.
+        let (tx, mut rx) = mpsc::unbounded::<Result<(),()>>();
+        let tx_clone = tx.clone();
+        self.set_on_close(move |_| {
+            // Note [mwu] Ignore argument, `CloseEvent` here contains rubbish
+            // anyway, nothing useful to pass to caller. Error code or reason
+            // string should not be relied upon.
+            utils::channel::emit(&tx_clone,Err(()));
+        });
+        self.set_on_open(move |_| {
+            utils::channel::emit(&tx,Ok(()));
+        });
+
+        match rx.next().await {
+            Some(Ok(())) => {
+                self.clear_callbacka();
+                Ok(())
+            }
+            _ => Err(ConnectingError::FailedToConnect)
+        }
+    }
+
+    /// Checks the current state of the connection.
+    pub fn state(&self) -> State {
+        State::query_ws(&self.ws)
+    }
+
+    /// Sets callback for the `close` event.
+    pub fn set_on_close(&mut self, f:impl FnMut(CloseEvent) + 'static) {
+        self.on_close.wrap(f);
+        self.ws.set_onclose(self.on_close.js_ref());
+    }
+
+    /// Sets callback for the `error` event.
+    pub fn set_on_error(&mut self, f:impl FnMut(Event) + 'static) {
+        self.on_error.wrap(f);
+        self.ws.set_onerror(self.on_error.js_ref());
+    }
+
+    /// Sets callback for the `message` event.
+    pub fn set_on_message(&mut self, f:impl FnMut(MessageEvent) + 'static) {
+        self.on_message.wrap(f);
+        self.ws.set_onmessage(self.on_message.js_ref());
+    }
+
+    /// Sets callback for the `open` event.
+    pub fn set_on_open(&mut self, f:impl FnMut(Event) + 'static) {
+        self.on_open.wrap(f);
+        self.ws.set_onopen(self.on_open.js_ref());
+    }
+
+    /// Clears all the available callbacks.
+    pub fn clear_callbacka(&mut self) {
```

Comment:
I believe there is a type in this name, unless it is a polish word for "callback" - "ustaw callbacka" :D

Reply:
done